sycl/doc/Assert.md

```diff
@@ -0,0 +1,421 @@
+assertion. The assertion failure message is printed to `stderr` by DPCPP
+Runtime.
+
+When multiple kernels are enqueued and more than one fail at assertion, at least
+single assertion should be reported.
+
+
+## User requirements
+
+From user's point of view there are the following requirements:
+
+| # | Title | Description | Importance |
+| - | ----- | ----------- | ---------- |
+| 1 | Abort DPC++ application | Abort host application when assert function is called and print a message about assertion | Must have |
+| 2 | Print assert message | Assert function should print message to stderr at host | Must have |
+| 3 | Stop under debugger | When debugger is attached, break at assertion point | Highly desired |
+| 4 | Reliability | Assert failure should be reported regardless of kernel deadlock | Highly desired |
+
+Implementations without enough capabilities to implement fourth requirement are
+allowed to realize the fallback approach described below, which does not
+guarantee assertion failure delivery to host, but is still useful in many
+practical cases.
+
+
+## Terms
+
+ - Device-side Runtime - runtime library supplied by the Native Device Compiler
+   and running on the device.
+ - Native Device Compiler - compiler which generates device-native binary image
+   based on input SPIR-V image.
+ - Low-level Runtime - the backend/runtime behind DPCPP Runtime attached via the
+   Plugin Interface.
+
+
+## How it works?
+
+`assert(expr)` macro ends up in call to `__devicelib_assert_fail`. This function
+is part of [Device library extension](extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst#cl_intel_devicelib_cassert).
+
+The format of the assert message is unspecified, but it will always include the
+text of the failing expression, the values of the standard macros `__FILE__` and
+`__LINE__`, and the value of the standard variable `__func__`. If the failing
+assert comes from an `nd_range` `parallel_for` it will also include the global
+ID and the local ID of the failing work item.
+
+Implementation of this function is supplied by Native Device Compiler for
+safe approach or by DPCPP Compiler for fallback one.
+
+In order to distinguish which implementation to use, DPCPP Runtime checks for
+`cl_intel_devicelib_cassert` extension. If the extension isn't available, then
+fallback implementation is used.
+
+
+## Safe approach
+
+This is the preferred approach and implementations should use it when possible.
+It guarantees assertion failure notification delivery to the host regardless of
+kernel behavior which hit the assertion.
+
+The Native Device Compiler is responsible for providing implementation of
+`__devicelib_assert_fail` which completely hides details of communication
+between the device code and the Low-Level Runtime from the SYCL device compiler
+and runtime. The Low-Level Runtime is responsible for:
+ - detecting if assert failure took place;
+ - flushing assert message to `stderr` on host.
```

Comment:
The intro paragraph contradicts to the last bullet:

> The assertion failure message is printed to `stderr` by DPCPP Runtime.

Reply:
Fixed